packages/astro/src/core/sync/write-files.ts

```diff
@@ -0,0 +1,88 @@
+import type fsMod from 'node:fs';
+import { dirname, relative } from 'node:path';
+import { fileURLToPath } from 'node:url';
+import { bold } from 'kleur/colors';
+import { normalizePath } from 'vite';
+import type { AstroSettings } from '../../@types/astro.js';
+import type { Logger } from '../logger/core.js';
+import { REFERENCE_FILE } from './constants.js';
+import { parse } from '@babel/parser';
+import generate from '@babel/generator';
+
+export async function writeFiles(settings: AstroSettings, fs: typeof fsMod, logger: Logger) {
+	writeInjectedTypes(settings, fs);
+	await setUpEnvTs(settings, fs, logger);
+}
+
+function getTsReference(type: 'path' | 'types', value: string) {
+	return `/// <reference ${type}=${JSON.stringify(value)} />`;
+}
+
+const CLIENT_TYPES_REFERENCE = getTsReference('types', 'astro/client');
+
+function writeInjectedTypes(settings: AstroSettings, fs: typeof fsMod) {
+	const references: Array<string> = [];
+
+	for (const { filename, content } of settings.injectedTypes) {
+		const filepath = normalizePath(fileURLToPath(new URL(filename, settings.dotAstroDir)));
+		fs.mkdirSync(dirname(filepath), { recursive: true });
+		fs.writeFileSync(filepath, formatContent(content), 'utf-8');
+		references.push(normalizePath(relative(fileURLToPath(settings.dotAstroDir), filepath)));
+	}
+
+	const astroDtsContent = `${CLIENT_TYPES_REFERENCE}\n${references.map((reference) => getTsReference('path', reference)).join('\n')}`;
+	if (references.length === 0) {
+		fs.mkdirSync(settings.dotAstroDir, { recursive: true });
+	}
+	fs.writeFileSync(
+		normalizePath(fileURLToPath(new URL(REFERENCE_FILE, settings.dotAstroDir))),
+		astroDtsContent,
+		'utf-8'
+	);
+}
+
+async function setUpEnvTs(settings: AstroSettings, fs: typeof fsMod, logger: Logger) {
+	const envTsPath = normalizePath(fileURLToPath(new URL('env.d.ts', settings.config.srcDir)));
```

Comment:
We should have a big `try/catch` block to catch all the possible file system errors that could be thrown

Reply:
Where should such try/catch be? maybe wrap everything inside `writeFiles`?

Reply:
The entire body of the function `setUpEnvTs`, because I see various FS operations here and in other non-exported function. In the end, the reason is the same, so we can handle all of them with only one single try/catch block